### PR TITLE
chore(refactor-polling): refactor polling rebased

### PR DESCRIPTION
**Pull request by anxolin**
# Summary
Refactor polling + Poll only if window Visible

Recreates https://github.com/cowprotocol/cowswap/pull/1967

Iterates over PR https://github.com/cowprotocol/cowswap/pull/1913 to refactor a bunch of places where we use polling. 

This PR creates a new hook: `useInterval`

This hook will periodically invoke a callback function and ensure the polling is not done while the browser tab/window is not visible.

## Motivation
While addressing this comment from Leandro https://github.com/cowprotocol/cowswap/pull/1913#pullrequestreview-1267805562 I saw how many repetitions we have in the polling logic. So, I gave it a try to use refactor into a new hook. 

## Additionally (unrelated to PR)
In this PR i was seeing a bunch of re-rendering issues during my tests, so you will see some changes:
- I added some memorization to improve the re-renders on some components that were re-rendering too much
- I renamed the component name on many updaters, because React use the function name to report errors, and there many components called `Updater`. I just gave them a name

## Testing
Check the updates in the console
image

This PR changes all the updaters and all logics that run something at intervals, so probably we should keep an eye that no flow is broken.

### CODE REVIEW

File: src/legacy/components/OrderProgressBar/index.tsx [hunk 1]
```diff
@@ -92,18 +93,20 @@ export function OrderProgressBar(props: OrderProgressBarProps) {
...
-    const id = setInterval(() => {
-      const percentage = getPercentage(elapsedSeconds, expirationInSeconds, chainId)
-      setPercentage(percentage)
-    }, REFRESH_INTERVAL_MS)
+    const percentage = getPercentage(elapsedSeconds, expirationInSeconds, chainId)
+    setPercentage(percentage)
+  }, [chainId, elapsedSeconds, expirationInSeconds, isPending])
 
-    return () => clearInterval(id)
-  }, [creationTime, validTo, chainId, elapsedSeconds, expirationInSeconds, isPending])
```
Comment (anxolin) [inline src/legacy/components/OrderProgressBar/index.tsx:106]: Anyone knows why `creationTime` was part of the useEffect? 

If we need to trigger it under some field changes we can, but  I'm not sure this is what we want. We already update 5 times per second!

`useEffect(updatePercentage, [creationTime, validTo, updatePercentage])`

Comment (alfetopito) [reply src/legacy/components/OrderProgressBar/index.tsx:106]: Can you use git blame and check when it's added?
At first glance I'd guess it was used in the `useEffect` and later removed, but the deps array was not updated

File: src/common/hooks/useInterval.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { useEffect, useRef } from 'react'
+
+import useIsWindowVisible from 'legacy/hooks/useIsWindowVisible'
+
+export interface UseIntervalParams {
+  /**
+   * Callback to be executed on each interval
+   * @returns a
```
Comment (alfetopito) [inline src/common/hooks/useInterval.ts:8]: Returns what?

File: src/common/hooks/useInterval.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  /**
+   * If true, the callback will be executed immediately instead of waiting for the next interval
+   */
+  triggerEagerly: boolean
+
+  /**
+   * Name of the polling function. Just for debugging porpouses
+   */
+  name: string
```
Comment (alfetopito) [inline src/common/hooks/useInterval.ts:25]: If just for debugging, can it be made optional?

File: src/common/hooks/useInterval.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+/**
+ * Invokes callback repeatedly over an interval defined by the delay
+ */
+export function useInterval(props: UseIntervalParams) {
+  const { callback, delay, triggerEagerly = true, name } = props
+  const isWindowVisible = useIsWindowVisible()
+  const savedCallback = useRef<() => void>()
```
Comment (alfetopito) [inline src/common/hooks/useInterval.ts:34]: Why not start the `useRef` with the callback in the constructor?

```suggestion
  const savedCallback = useRef<() => void>(callback)
```

File: src/common/hooks/useInterval.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  // Remember the latest callback.
+  useEffect(() => {
+    savedCallback.current = callback
+  }, [callback])
+
+  // Set up the interval.
+  useEffect(() => {
+    function executeCallBack() {
+      const { current: currentCallback } = savedCallback
+      if (isWindowVisible && currentCallback) {
```
Comment (alfetopito) [inline src/common/hooks/useInterval.ts:45]: will `current` ever be falsy? (I mean, if you pass it in the constructor it shouldn't)

File: src/common/hooks/useInterval.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (delay === null || !isWindowVisible) {
+      return
+    }
+
+    console.debug(`[useInterval:${name}] Schedule interval`)
+    if (triggerEagerly) {
+      executeCallBack()
+    }
+    const id = setInterval(executeCallBack, delay)
```
Comment (alfetopito) [inline src/common/hooks/useInterval.ts:59]: What if `delay` is set to `0`?

**Review by elena-zh — APPROVED** [review]

File: src/legacy/state/lists/updater.ts [hunk 7]
```diff
@@ -35,7 +35,12 @@ export default function Updater(): null {
   }, [fetchList, isWindowVisible, lists])
 
   // fetch all lists every 10 minutes, but only after we initialize provider
-  useInterval(fetchAllListsCallback, provider ? 1000 * 60 * 10 : null)
+  useInterval({
+    callback: fetchAllListsCallback,
+    name: 'ListsUpdater',
+    delay: provider ? 1000 * 5 : null, // provider ? ms`10m` : null,
```
Comment (alfetopito) [inline src/legacy/state/lists/updater.ts:41]: This is changing the poll frequency.
Is it intentional to go from 10min to 5 seconds?

File: src/legacy/state/price/updater.ts [hunk 8]
```diff
@@ -275,5 +261,18 @@ export default function FeesUpdater(): null {
...
   ])
 
+  /**
+   * Periodically fetch quotes
+   */
+  useInterval({
+    callback: refetchQuoteIfRequired,
+    name: 'FeesUpdater',
+    delay: REFETCH_CHECK_INTERVAL,
+    triggerEagerly: true,
```
Comment (alfetopito) [inline src/legacy/state/price/updater.ts:271]: Should this be `false` given that there's an `useEffect` below triggering when the callback changes?

```suggestion
    triggerEagerly: false,
```

File: src/modules/fortune/containers/FortuneWidget/index.tsx [hunk 9]
```diff
@@ -296,7 +297,14 @@ export function FortuneWidget() {
   const [isFortunedShared, setIsFortunedShared] = useState(false)
 
   const [today, setToday] = useState(new Date())
-  useInterval(() => setToday(new Date()), 2_000)
+  const updateToday = useCallback(() => setToday(new Date()), [])
+  useInterval({
+    callback: updateToday,
+    name: 'FortuneWidget',
+    delay: ms`10m`,
```
Comment (alfetopito) [inline src/modules/fortune/containers/FortuneWidget/index.tsx:304]: Moving update from 2 seconds to 10min intentional?

Comment (anxolin) [reply src/modules/fortune/containers/FortuneWidget/index.tsx:304]: I should have mention it. I was seeing a lot of checks on this, but i didn't see what is the point on doing this so frequently. 

@shoom3301 is not 10min good enough?

File: src/legacy/state/orders/updaters/CancelledOrdersUpdater.ts [hunk 10]
```diff
@@ -39,94 +39,94 @@ export function CancelledOrdersUpdater(): null {
...
-  const updateOrders = useCallback(
-    async (chainId: ChainId, account: string) => {
-      const lowerCaseAccount = account.toLowerCase()
-      const now = Date.now()
-
-      if (isUpdating.current) {
-        return
-      }
+  const updateOrders = useCallback(async () => {
+    if (!chainId || !account || isUpdating.current) {
```
Comment (shoom3301) [inline src/legacy/state/orders/updaters/CancelledOrdersUpdater.ts:43]: `chainId` can't be falsy, no need to check it

File: src/legacy/state/orders/updaters/CancelledOrdersUpdater.ts [hunk 11]
```diff
@@ -39,94 +39,94 @@ export function CancelledOrdersUpdater(): null {
...
+    } finally {
+      isUpdating.current = false
+      // console.debug(`[CancelledOrdersUpdater] Checked recently canceled orders in ${Date.now() - startTime}ms`)
     }
+  }, [addOrderToSurplusQueue, fulfillOrdersBatch, account, chainId])
 
-    const interval = setInterval(() => updateOrders(chainId, account), MARKET_OPERATOR_API_POLL_INTERVAL)
-
-    return () => clearInterval(interval)
+  useEffect(() => {
```
Comment (shoom3301) [inline src/legacy/state/orders/updaters/CancelledOrdersUpdater.ts:120]: Should we do the same initial `useEffect` for other orders updaters?

File: src/modules/fortune/containers/FortuneWidget/index.tsx [hunk 12]
```diff
@@ -306,6 +314,7 @@ export function FortuneWidget() {
     : ''
 
   const isDailyFortuneChecked = useMemo(() => {
+    console.log('FortuneWidget', { lastCheckedFortune, today: today.toISOString() })
```
Comment (shoom3301) [inline src/modules/fortune/containers/FortuneWidget/index.tsx:317]: Please, don't forget to remove this

**Pull request closed without merging**